Sources/SourceControl/RepositoryManager.swift

```diff
@@ -165,54 +121,74 @@ public class RepositoryManager {
     /// Get a handle to a repository.
     ///
     /// This will initiate a clone of the repository automatically, if
-    /// necessary, and immediately return. The client can add observers to the
-    /// result in order to know when the repository is available.
-    public func lookup(repository: RepositorySpecifier) -> RepositoryHandle {
-        // Check to see if the repository has been provided.
-        if let handle = repositories[repository.url] {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (Result<RepositoryHandle, AnyError>) -> Void) {
+        DispatchQueue.global().async {
+            repostoriesLock.whileLocked {
+                // Check to see if the repository has been provided.
+                if let handle = self.repositories[repository.url] {
+                    // Fetch and update the repository when it is being looked up.
+                    // We need some mechanism to maybe not do this for pinned repos.
+                    do {
+                        let repo = try handle.open()
```

Comment:
This is holding the `repositoriesLock` while doing the expensive operation, which would prevent it from every being particularly concurrent.

This is one reason we need something like the handle to be able to be placed into the repositories mapping _while_ it is being fetched (since if someone else fetches it they need to just chain onto that same fetch operation). We might be able to foist this onto the client and say "don't clone the same URL multiple times" with this API, but someone has to handle managing this, and it seems ideal to keep it contained.